lib/services/expiry/expiry.go

```diff
@@ -0,0 +1,250 @@
+	s := &Service{
+		Config: cfg,
+		ctx:    ctx,
+	}
+	return s, nil
+}
+
+// Run starts the expiry service.
+func (s *Service) Run() error {
+	semCfg := services.SemaphoreLockConfigWithRetry{
+		SemaphoreLockConfig: services.SemaphoreLockConfig{
+			Service: s.AccessPoint,
+			Params: types.AcquireSemaphoreRequest{
+				SemaphoreKind: types.KindAccessRequest,
+				SemaphoreName: semaphoreName,
+				MaxLeases:     1,
+				Holder:        s.HostID,
+			},
+			Expiry: semaphoreExpiration,
+			Clock:  s.Clock,
+		},
+		Retry: retryutils.LinearConfig{
+			Clock:  s.Clock,
+			First:  time.Second,
+			Step:   semaphoreExpiration / 2,
+			Max:    semaphoreExpiration,
+			Jitter: retryutils.DefaultJitter,
+		},
+	}
+
+	poll := interval.New(interval.Config{
+		Duration:      scanInterval,
+		FirstDuration: retryutils.FullJitter(scanInterval),
+		Jitter:        retryutils.SeventhJitter,
+		Clock:         s.Clock,
+	})
+	defer poll.Stop()
+
+	for {
+		lease, err := services.AcquireSemaphoreLockWithRetry(
+			s.ctx,
+			semCfg,
+		)
+		if err != nil {
+			s.Log.WarnContext(s.ctx, "error aquiring semaphore", "error", err)
+			continue
+		}
+
+		if err := s.processRequests(); err != nil {
+			s.Log.WarnContext(s.ctx, "error processing access requests", "error", err)
+		}
+		ctx, cancel := context.WithCancel(lease)
+		defer cancel()
+		lease.Stop()
+		if err := lease.Wait(); err != nil {
+			s.Log.WarnContext(ctx, "error cleaning up semaphore", "error", err)
+		}
+		select {
+		case <-s.ctx.Done():
+			return nil
+		case <-poll.Next():
+		}
+	}
+}
+func (s *Service) processRequests() error {
+	requestsExpired := 0
+	nextPageStart := ""
+	for {
+		var page []*types.AccessRequestV3
+		var err error
+		readTime := s.Clock.Now()
+		page, nextPageStart, err = s.getNextPageOfAccessRequests(nextPageStart)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if len(page) == 0 || requestsExpired >= maxExpiresPerCycle {
+			return nil
+		}
+		minPageDelay := time.After(retryutils.SeventhJitter(minPageDelay))
+		for _, req := range page {
+			if !s.shouldExpire(req, readTime) {
+				continue
+			}
+			if err := s.expireRequest(s.ctx, req); err != nil {
+				return trace.Wrap(err)
+			}
+			requestsExpired++
+			if requestsExpired >= maxExpiresPerCycle {
+				return nil
+			}
+		}
+		<-minPageDelay
+	}
+}
+
+func (s *Service) getNextPageOfAccessRequests(startKey string) ([]*types.AccessRequestV3, string, error) {
+	req := &proto.ListAccessRequestsRequest{
+		Sort:       proto.AccessRequestSort_CREATED,
+		Descending: true,
+		Limit:      accessRequestPageLimit,
+		StartKey:   startKey,
+	}
+	resp, err := s.AccessPoint.ListAccessRequests(s.ctx, req)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+
+	return resp.AccessRequests, resp.NextKey, nil
+}
+
+func (s *Service) shouldExpire(req types.AccessRequest, readTime time.Time) bool {
+	expires := req.Expiry()
+	if req.GetState() == types.RequestState_PENDING {
+		expires = expires.Add(pendingRequestGracePeriod)
+	}
```

Comment:
Can you explain why this block is needed? It's not evident at first glance. A comment would be helpful too.

Reply:
This was added since the expiry of pending requests is extended when approved. The grace period was suggested as a solution [here](https://github.com/gravitational/teleport/pull/49267#issuecomment-2512383026)